src/main/java/org/opensearch/neuralsearch/query/NeuralSparseQuery.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+
+/**
+ * Implementation of Query interface for type "hybrid". It allows execution of multiple sub-queries and collect individual
+ * scores for each sub-query.
+ */
+@AllArgsConstructor
+@Getter
+@NonNull
+public final class NeuralSparseQuery extends Query {
+
+    private Query currentQuery;
+    private final Query highScoreTokenQuery;
+    private final Query lowScoreTokenQuery;
+    private final Float rescoreWindowSizeExpansion;
+
+    @Override
+    public String toString(String field) {
+        return "NeuralSparseQuery("
+            + currentQuery.toString(field)
+            + ","
+            + highScoreTokenQuery.toString(field)
+            + ", "
+            + lowScoreTokenQuery.toString(field)
+            + ")";
+
+    }
+
+    public Query rewrite(IndexSearcher indexSearcher) throws IOException {
+        Query rewrittenCurrentQuery = currentQuery.rewrite(indexSearcher);
+        Query rewrittenFirstStepQuery = highScoreTokenQuery.rewrite(indexSearcher);
+        Query rewrittenSecondPhaseQuery = lowScoreTokenQuery.rewrite(indexSearcher);
+        if (rewrittenFirstStepQuery == highScoreTokenQuery && rewrittenSecondPhaseQuery == lowScoreTokenQuery) {
+            return this;
+        }
+        return new NeuralSparseQuery(rewrittenCurrentQuery, rewrittenFirstStepQuery, rewrittenSecondPhaseQuery, rescoreWindowSizeExpansion);
+    }
+
+    @Override
+    public void visit(QueryVisitor queryVisitor) {
+        QueryVisitor v = queryVisitor.getSubVisitor(BooleanClause.Occur.SHOULD, this);
+        currentQuery.visit(v);
+        highScoreTokenQuery.visit(v);
+        lowScoreTokenQuery.visit(v);
+    }
```

Comment:
why we need this?

Reply:
I made a update.
Basically, each custom Query which extends Query need to implement this abstract to visit all child queries.
But at now, I think we can visit the current query only, because other sub queries in NeuralSparseQuery are subset of the current query at first.
It's happen on the DefaultAggregationProcessor.proProcess() and have no influence on the search progress now. So just visit the currentQuery will keep the same logic as this update before.
As we concerned, after building a interface on opensearch core we can remove this custom query class also.